web/ui/react-app/src/ExpressionInput.tsx

```diff
@@ -0,0 +1,157 @@
+import React, { Component } from 'react';
+import {
+  Button,
+  InputGroup,
+  InputGroupAddon,
+  InputGroupText,
+  Input,
+} from 'reactstrap';
+
+import Downshift, { ChildrenFunction, ControllerStateAndHelpers, DownshiftInterface } from 'downshift';
+import fuzzy from 'fuzzy';
+
+import { library } from '@fortawesome/fontawesome-svg-core';
+import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
+import { faSearch, faSpinner } from '@fortawesome/free-solid-svg-icons';
+
+library.add(faSearch, faSpinner);
+
+interface ExpressionInputProps {
+  value: string;
+  metricNames: string[];
+  executeQuery: (expr: string) => void;
+  loading: boolean;
+}
+
+class ExpressionInput extends Component<ExpressionInputProps> {
+  prevNoMatchValue: string | null = null;
+  private exprInputRef = React.createRef<HTMLInputElement>();
+
+  handleKeyPress = (event: React.KeyboardEvent<HTMLInputElement>) => {
+    if (event.key === 'Enter' && !event.shiftKey) {
+      this.props.executeQuery(this.exprInputRef.current!.value);
+      event.preventDefault();
+    }
+  }
+
+  renderAutosuggest = (downshift: any) => {
+    if (!downshift.isOpen) {
+      return null;
+    }
+
+    if (this.prevNoMatchValue && downshift.inputValue.includes(this.prevNoMatchValue)) {
+      return null;
+    }
+
+    let matches = fuzzy.filter(downshift.inputValue.replace(/ /g, ''), this.props.metricNames, {
+      pre: "<strong>",
+      post: "</strong>",
+    });
+
+    if (matches.length === 0) {
+      this.prevNoMatchValue = downshift.inputValue;
+      return null;
+    }
+
+    return (
+      <ul className="autosuggest-dropdown" {...downshift.getMenuProps()}>
+        {
+          matches
+            .slice(0, 200) // Limit DOM rendering to 100 results, as DOM rendering is sloooow.
+            .map((item, index) => (
+              <li
+                {...downshift.getItemProps({
+                  key: item.original,
+                  index,
+                  item: item.original,
+                  style: {
+                    backgroundColor:
+                      downshift.highlightedIndex === index ? 'lightgray' : 'white',
+                    fontWeight: downshift.selectedItem === item ? 'bold' : 'normal',
+                  },
+                })}
+              >
+                {/* TODO: Find better way than setting inner HTML dangerously. We just want the <strong> to not be escaped.
+                    This will be a problem when we save history and the user enters HTML into a query. */}
+                <span dangerouslySetInnerHTML={{__html: item.string}}></span>
+              </li>
+            ))
+        }
+      </ul>
+    );
+  }
+
+  componentDidMount() {
+    const $exprInput = $(this.exprInputRef.current!);
+    $exprInput.on('input', () => {
+      const el = $exprInput.get(0);
+      const offset = el.offsetHeight - el.clientHeight;
+      $exprInput.css('height', 'auto').css('height', el.scrollHeight + offset);
+    });
+  }
+
```

Comment:
I think height should be set on DidMount as well since now on page refresh Input is taking its initial size and text remains hidden. Btw, this can be replaced very easily with
`() => this.setState({ height: 'auto' }, () => this.setState({ height: this.inputRef.current.scrollHeight }))` attached as onInput handler and height as style attr

Reply:
Yep, it works fine but overall is better to delegate this job to react since another issue appears here. If the user clears the input (which contains multiple lines) and then the input loses its focus the query will appear again without input to be resized. This is happening because Downshift set its selectedItem with the value coming from this.props which in this case never changes. So, this can be resolved by making the input controlled and value to be saved in the local state together with the height of the input. I can make PR for this if you don't have mind :)
Also I was wonder whether redux can be integrated in the App. I'm not sure how much is going to grow but for sure will simplify the components and later on will be easier to extend end test

Reply:
@boyskila Yeah, it would be great to get a PR for improving this :) Getting Downshift to work just the way I wanted required a bit of hackery and dirtiness, and I hope there's better / cleaner / more efficient ways of doing it.

As for Redux, I've heard different opinions from React people, so at least I would like to consult more people before introducing it.

Reply:
@juliusv this PR https://github.com/prometheus/prometheus/pull/6174 is for the ExpressionInput clearing issue on blur.